### PR TITLE
Add hours

### PR DESCRIPTION
**Pull request by Oneill38**
-Added a calendar via rb script in /lib
-Changed calendars name to timetables to be more descriptive
-Hours are updated via ajax call, with plain SQL that uses a bulk upsert operation which is approx 83% faster than an iterative upsert, and 97% faster than a basic iterative approach. 
-Validations occur in the batch_update method since we're using straight sql w/a batch approach and can't check every entry one by one

### CODE REVIEW

**Comment by Oneill38** [comment]
Working on updating the method for batch update and the time validation to accept the tbd,closed, and note field

File: README.md [hunk 1]
```diff
@@ -3,6 +3,18 @@
 [![Build Status](https://travis-ci.org/cul/ldpd-hours.svg?branch=master)](https://travis-ci.org/cul/ldpd-hours)
 [![Code Climate](https://codeclimate.com/github/cul/ldpd-hours/badges/gpa.svg)](https://codeclimate.com/github/cul/ldpd-hours)
 [![Issue Count](https://codeclimate.com/github/cul/ldpd-hours/badges/issue_count.svg)](https://codeclimate.com/github/cul/ldpd-hours)
+
+## Installation
```
Comment (cgalarza) [inline README.md]: Specify that this is for development environments?

File: .gitignore [hunk 2]
```diff
@@ -21,3 +21,5 @@
 /yarn-error.log
 
 .byebug_history
+
+/config/database.yml
```
Comment (cgalarza) [inline .gitignore:25]: It's helpful to have something like a `config/database.template.yml` with example configuration for developers setting up a new development instance. I would add one in.

File: app/controllers/timetables_controller.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
+class TimetablesController < ApplicationController
+  load_and_authorize_resource
+
+  def batch_edit
+		@library = Library.find(params["library_id"])
+		@date = params[:date] ? Date.parse(params[:date]) : Date.today
+		@timetable = Timetable.new
+  end
+
+  def batch_update
```
Comment (cgalarza) [inline app/controllers/timetables_controller.rb:10]: Might be helpful to use strong params here, just to clearly list what parameters are required/permitted.

File: app/controllers/timetables_controller.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  private
+
+  def format_dates(dates)
+		dates.map!{|selected_date| Date.parse(selected_date) }
+  end
+
+  def opens_before_close(open,close, closed, tbd)
+    return true if (closed == "1" || tbd == "1")
+		if !(Time.parse(open, Time.now) < Time.parse(close, Time.now))  
+			raise StandardError
```
Comment (cgalarza) [inline app/controllers/timetables_controller.rb]: Maybe an ArgumentError would be better here? Up to you, either way we should add a description something like `raise StandardError, "End time cannot be before start time"`

File: app/controllers/timetables_controller.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  private
+
+  def format_dates(dates)
+		dates.map!{|selected_date| Date.parse(selected_date) }
+  end
+
+  def opens_before_close(open,close, closed, tbd)
+    return true if (closed == "1" || tbd == "1")
```
Comment (cgalarza) [inline app/controllers/timetables_controller.rb]: I would combine this to one if/else tree, i think it would make it easier to follow the logic.

File: app/models/timetable.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+class Timetable < ApplicationRecord
+	belongs_to :library
+
+	def self.batch_update_or_create(library_id, dates, open, close, closed, tbd, note)
+		vals = dates.map!{|day| "('#{library_id}', '#{day}', '#{open}', '#{close}', #{closed}, #{tbd}, '#{note}', NOW(), NOW())"}.join(", ")
```
Comment (cgalarza) [inline app/models/timetable.rb]: Because this is writing straight sql with parameters provided by the user, we really need to validate the parameters within this method and sanitize the query we create or use interpolation to create the query. Some of these checks are done in the controller, but they should happen again closer to the creation of the sql query in case this method is used in a different controller that doesn't do the same cleanup.

File: config/application.rb [hunk 7]
```diff
@@ -11,6 +11,7 @@ class Application < Rails::Application
   	include Cul::Omniauth::FileConfigurable
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 5.1
+    config.autoload_paths += %W(#{config.root}/lib)
```
Comment (cgalarza) [inline config/application.rb:14]: Why was this needed?

File: config/routes.rb [hunk 8]
```diff
@@ -1,16 +1,28 @@
...
-	  get 'sign_in', :to => 'users/sessions#new', :as => :new_user_session
-	  get 'sign_out', :to => 'users/sessions#destroy', :as => :destroy_user_session
-	end
+  	  get 'sign_in', :to => 'users/sessions#new', :as => :new_user_session
+  	  get 'sign_out', :to => 'users/sessions#destroy', :as => :destroy_user_session
+  	end
+
+	resources :libraries do
+    resources :timetables do
+      collection do
```
Comment (cgalarza) [inline config/routes.rb:16]: nice! The ones commented out at the bottom can be deleted?

File: db/schema.rb [hunk 9]
```diff
@@ -23,7 +23,22 @@
...
+    t.string "open"
+    t.string "close"
+    t.text "notes"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.boolean "tbd", default: false
+    t.boolean "closed", default: false
+    t.string "note"
+    t.bigint "library_id"
+    t.index ["library_id", "date"], name: "index_timetables_on_library_id_and_date", unique: true
```
Comment (cgalarza) [inline db/schema.rb:37]: nice!

File: lib/calendar_builder.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
+class CalendarBuilder < Struct.new(:view, :date, :callback)
+    HEADER = %w[Sunday Monday Tuesday Wednesday Thursday Friday Saturday]
```
Comment (cgalarza) [inline lib/calendar_builder.rb:2]: I believe the Date class in Ruby has a constant with the day of the week, would be nice to use here.

File: spec/controllers/timetables_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+require 'rails_helper'
+
+RSpec.describe TimetablesController, type: :controller do
+
+  describe "#format_dates" do
+    it "returns an array of date objects" do
+      dates = ["2017-06-06", "2017-05-29"]
+      expect(TimetablesController.new.send(:format_dates, dates).first.class).to eql(Date)
```
Comment (cgalarza) [inline spec/controllers/timetables_controller_spec.rb]: I thought you could use a `controller` variable to get at the current controller in controller specs. If that's a thing i made up I would add a this to the top and use it through out `let(:controller) { TimetablesController.new }`

File: spec/features/libraries_spec.rb [hunk 12]
```diff
@@ -1,6 +1,12 @@
 require 'rails_helper'
 
 describe "Libraries" do
+  include_context 'login admin user'
+  
+  before(:each) do
+    @library = Library.create(name: "Lehman", code: "lehman")
```
Comment (cgalarza) [inline spec/features/libraries_spec.rb:7]: factory?

File: spec/features/timetables_spec.rb [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  before(:each) do
+    @library = Library.create(name: "Lehman", code: "lehman")
+  end
+
+  describe "viewing the page" do
+    include_context 'login admin user'
+
+    it "should display a calendar" do
+      visit(batch_edit_library_timetables_path(@library))
+      expect(page).to have_css("table")
```
Comment (cgalarza) [inline spec/features/timetables_spec.rb]: maybe a table with a more specific class?

File: spec/models/timetables_spec.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+		it "should add multiple dates" do
+			code, dates, open, close, closed, tbd, note = Library.first.id, [Date.today, Date.today + 1.month], "07:00AM", "08:00PM", false, false, ""
+			Timetable.batch_update_or_create(code,dates,open,close,closed,tbd,note)
+			expect(Timetable.count).to eq(2)
+		end
+
+		it "should not have an open or close time for a closed day" do
+			code, dates, open, close, closed, tbd, note = Library.first.id, [Date.today], nil, nil, true, false, ""
+			Timetable.batch_update_or_create(code,dates,open,close,closed,tbd,note)
```
Comment (cgalarza) [inline spec/models/timetables_spec.rb]: For this method, it might be good to use named parameters. Optional, up to you.

File: db/schema.rb [hunk 15]
```diff
@@ -23,7 +23,22 @@
     t.datetime "updated_at", null: false
   end
 
-  create_table "users", force: :cascade do |t|
+  create_table "timetables", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8" do |t|
+    t.date "date"
+    t.string "open"
```
Comment (cgalarza) [inline db/schema.rb:28]: I am not sure if this is worth changing now, but since we are using "closed" to represented whether or not a library is closed and "close" for the time it seems like those are a little too similar. Maybe open_time, close_time might make those columns a little clearer. Again, not sure if it's worth changing.

**Review by cgalarza — CHANGES_REQUESTED** [review]
The interface looks really, really nice! It works great. We'll have to provide some more instructions, but that can come later. One thing that we might want to add later is some javascript to disable options depending on what the user selects. For example, if the user has selected "closed", tbd and the time select boxes should be disabled.

File: spec/features/libraries_spec.rb (hunk 12 above)
Comment (cgalarza) [reply spec/features/libraries_spec.rb:7]: I just realized that we don't have Factory Girl in this project, yet. I'll add it to my branch and we can convert these to factories later on (I totally thought we already had the gem installed).

File: lib/calendar_builder.rb (hunk 10 above)
Comment (Oneill38) [reply lib/calendar_builder.rb:2]: Going to leave this one alone for now in interest of time/since the days won't ever change

File: config/application.rb (hunk 7 above)
Comment (Oneill38) [reply config/application.rb:14]: the calendar_builder is located in lib

**Comment by Oneill38** [comment]
Pushed up the requested changes, ready for rereview whenever

File: config/application.rb (hunk 7 above)
Comment (cgalarza) [reply config/application.rb:14]: Isn't /lib automatically part of the load path?

Comment (Oneill38) [reply config/application.rb:14]: Looks like it isn't anymore http://guides.rubyonrails.org/autoloading_and_reloading_constants.html

Comment (cgalarza) [reply config/application.rb:14]: Apparently I also have this set up in AC. I'll look into this more later, because I am curious. But we can leave it in for now.

File: spec/models/timetables_spec.rb [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+RSpec.describe :Timetable, type: :model do
+  before(:each) do
+    Library.create(name: "Butler", code: "butler")
+  end
+	describe "batch_update_or_create" do
+		it "should add records to the time_tables table" do
+			code, dates, open, close, closed, tbd, note = Library.first.id, [Date.today], "07:00AM", "08:00PM", false, false, ""
+			Timetable.batch_update_or_create({"library_id" => code, "dates" => dates, "closed" => closed, "tbd" => tbd, "note" => note}, open, close)
+			expect(Timetable.last.date).to eq(Date.today)
```
Comment (cgalarza) [inline spec/models/timetables_spec.rb:11]: Might be good to check that the time was set correctly?

**Comment by cgalarza** [comment]
Added one comment about adding a bit to testing. Otherwise looks great!

File: config/application.rb (hunk 7 above)
Comment (cgalarza) [reply config/application.rb:14]: @Oneill38 AHHHH, i missed your earlier comment, good to know!
